### ISSUE TITLE
Cannot verify hostname, who is to blame ? Logcat output

### ISSUE THREAD

**Issue by jbd** [bitfireAT/davdroid#332]
Hi,

I'm using Cyanogen Cyanogen 11 on my Moto G.

I've got the classic "Cannot verify hostname" on webmail.web4all.fr (I'm not owning the domain). The certificate looks OK to me (https://www.ssllabs.com/ssltest/analyze.html?d=webmail.web4all.fr). While roughly looking at the logcat output (https://gist.github.com/jbd/4efd6003f62562eb4b64), I'm not sure if the problem is on the server or client side. 

Some lines looks suspicious to me :

I/davdroid.WebDavResource( 8269): Redirection on PROPFIND; trying again at new content URL: https://webmail.web4all.fr:80/dav

What is your opinion ?

**Comment by nakedhitman** [comment]
I am also having this problem, but in my own private cloud. I am running Zimbra 8.5 behind an Nginx SSL reverse-proxy and have a StartSSL wildcard cert. If it would help in your testing, I can provide my configs and a test user account you can use to see what is going on.

**Comment by nakedhitman** [comment]
This may be the dumbest workaround ever, but I got a sync to start by making sslh listen on port 80, and sort traffic to my reverse proxy, which now listens on 8080 and 443. This is a not-insignificant waste of resources, since it spins off a thread for every connection. Is there any information I can provide to help get to the bottom of this?

**Comment by dper** [comment]
@nakedhitman If one reverse proxy fails (nginx) but another works (sslh), it's a good sign you have a reverse proxy problem.

The extra thread per connection probably won't affect your system performance at all.  If you're the only user, the extra system load should be negligible.

**Comment by nakedhitman** [comment]
@dper That's entirely possible, however all that sslh is doing is allowing both HTTP and HTTPS traffic over port 80 and is redirecting both kinds of traffic to the Nginx reverse-proxy. Something is still causing the HTTPS dav requests to be redirected to port 80, which makes little sense to me. 
If any of you have any Nginx skills, I have posted my config from before I employed sslh at http://pastebin.com/kFqXvVXc. Thanks!

**Comment by untitaker** [comment]
@nakedhitman In your case, maybe it's useful to know that nginx might block certain HTTP methods that are required for DAV, except when using PHP-DAV-servers via php-fpm.

E.g. for reverse-proxying to Radicale:

    location /radicale/ {
        dav_methods  PUT DELETE MKCOL COPY MOVE;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_pass   http://127.0.0.1:5232;
    }

**Comment by untitaker** [comment]
Also check out nginx' error log.

**Comment by dper** [comment]
@nakedhitman I would be happy to look at the pastebin, but PasteBin uses CloudFlare, and CloudFlare blocks Tor connections. :-( GitHub allows Tor connections, though. :-)

**Comment by untitaker** [comment]
Rehosting: http://sprunge.us/hdQa

**Comment by dper** [comment]
@untitaker Thanks!

**Comment by nakedhitman** [comment]
@untitaker, I added that dav line you suggested and reverted all other settings to what you see above. The only error message in my logs is this one, which hasn't caused any problems with the other services I'm running:
     [error] 26431#0: OCSP_basic_verify() failed (SSL: error:27069065:OCSP routines:OCSP_basic_verify:certificate verify error:Verify error:unable to get local issuer certificate) while requesting certificate status, responder: ocsp.startssl.com
If any of you are interested, I have created a test user account: davtest@zimbra.local/DAVTest123!

**Comment by untitaker** [comment]
I have the ``ssl_``-options inside the server clause, so maybe try that?. Also I find it a bit weird that you listen to two ports at the same server, but if it's working, so be it.

>If any of you are interested, I have created a test user account: davtest@zimbra.local/DAVTest123!

I don't think we'll be able to reach your installation from here :)

**Comment by nakedhitman** [comment]
You should be able to log in to mail.awesomegeek.com with that account from anywhere that allows traffic on port 443. I am currently off-site and can access it just fine. The SSL options are allowed to live just about anywhere, and are currently working for many other services with the above config. cloud.awesomegeek.com is currently running an OwnCloud instance behind the same reverse-proxy with another server section that I have not included for brevity. I have been successfully able to mount WebDAV shares from it, so I know that at least some DAV is able to work through this proxy config. If you want my full server list, I can provide that, but it didn't seem relevant.

**Comment by untitaker** [comment]
Ah yes, that's the information we're missing :P

But seriously, I'll give it a try tomorrow or so.

**Comment by dper** [comment]
I notice you're using StartSSL.  To make nginx and StartSSL work together properly, there's some certificate concatenation that's needed.  Have you done so?  If not, this could explain the above error.  (If you don't properly concatenate the certificates, some things will work and others won't, which makes it hard to debug.)

**Comment by nakedhitman** [comment]
@dper Yes, I have concatenated the certs. Validation gets an A rating from https://www.ssllabs.com/ssltest/analyze.html?d=awesomegeek.com. 

I just created a davtest account on my OwnCloud instance and have successfully gotten DAVdroid to sync to it without issues. I'm beginning to suspect that this is more of a Zimbra issue, so I will be opening a thread on their forums. Thanks for looking at this, guys. Hopefully we can get to a solution that will benefit other users. I'll post back if a fix is identified.
